### ISSUE TITLE
Add `cargoTest` task

### ISSUE THREAD

**Issue by dcsommer** [mozilla/rust-android-gradle#67]
It might make sense to allow running `cargo test` via a Gradle task. My particular use case requires this because
- My Rust code depends on static libraries built in another gradle task
- My build.rs relies on extra environment variables provided injected via rust-android-gradle's `exec` option to compile properly
- The repo's test runner is configured to interface with gradle already, so it makes sense to add the Rust tests there too

**Comment by ncalexan** [comment]
Yep.  I feel like we had a ticket about this, but now I don't see one.  Perhaps it was in the old ncalexan/rust-android-gradle repo and didn't migrate.

I'd gladly take a patch for this; it shouldn't be too hard to arrange.  Bonus points if it can be attached to the `check` Gradle target (perhaps manually, at first).

As follow-up, it would be nice to assert this works by building on top of the https://github.com/ncalexan/rust-android-gradle/tree/citest branch (which I do mean to finish up at some point, but work and life don't leave much time these days).

**Comment by dcsommer** [comment]
Relevant to this issue, there is a question of supporting arbitrary cargo commands. Many `cargo` CLI workflows require the environment variables set by this plugin. Should we consider adding a way to specify all these workflows as part of the plugin definition? In the end, tasks could be as complicated as `cargo${commandKey}${buildType}${productFlavor}${ABI}`, which is getting pretty long.

Hypothetically you could have something like this in the `cargo` block:

```
cargo {
  commands = [
    "normalBuild": ["build"],
    "buildTests": ["build", "--tests"],
    "runTests": ["test", "--all"],
    "clippy": ["clippy"],
    "fmt": ["fmt"],
  ]
}
```

Some of the existing options like `extraCargoBuildArguments` could be deprecated given the above API.

One could argue that `cargo ndk` is the right way to generalize these workflows, but I have no way to stably derive the state I pass into `rust-android-gradle` outside Gradle itself. In particular, for mixed C/C++/Rust builds, getting things like output paths for CMake builds feels like it should stay in `build.gradle`. Replicating the way the Android Gradle Plugin invokes CMake in a build.rs seems less reliable.

**Comment by ncalexan** [comment]
> Relevant to this issue, there is a question of supporting arbitrary cargo commands. Many `cargo` CLI workflows require the environment variables set by this plugin. Should we consider adding a way to specify all these workflows as part of the plugin definition? In the end, tasks could be as complicated as `cargo${commandKey}${buildType}${productFlavor}${ABI}`, which is getting pretty long.
> 
> Hypothetically you could have something like this in the `cargo` block:
> 
> ```
> cargo {
>   commands = [
>     "normalBuild": ["build"],
>     "buildTests": ["build", "--tests"],
>     "runTests": ["test", "--all"],
>     "clippy": ["clippy"],
>     "fmt": ["fmt"],
>   ]
> }
> ```

I would be happy to explore something in this space.  My preference would be more like:
```
cargo {
    commands {
        buildTests { spec, toolchain, androidVariant, commandName ->
            // Some configuration.
            spec.args(["build", "--tests"])
            spec.environment("TEST", "1")
        }
        ...
    }
```
But we can bikeshed on the details at some point in the future.

In general, I am happy to turn `rust-android-gradle` into a glue layer between Gradle and invoking `cargo` in various ways.  Provided, of course, that we keep the main functionality (build, maybe check, maybe test) supported by default, preferably with no configuration required.

> Some of the existing options like `extraCargoBuildArguments` could be deprecated given the above API.

Yes.  And the existing `exec`, which is a wart-y escape hatch, would be put into a greater context:
```
cargo {
    commands {
        build {
            // This is where `exec` goes now.
        }
    }
}
```

> One could argue that `cargo ndk` is the right way to generalize these workflows, but I have no way to stably derive the state I pass into `rust-android-gradle` outside Gradle itself. In particular, for mixed C/C++/Rust builds, getting things like output paths for CMake builds feels like it should stay in `build.gradle`. Replicating the way the Android Gradle Plugin invokes CMake in a build.rs seems less reliable.

I can't speak to this challenge.  Can you link to your project that's bringing together CMake and `rust-android-gradle` open source, so that I can see what you're trying to achieve?

**Comment by dcsommer** [comment]
> Can you link to your project that's bringing together CMake and rust-android-gradle open source, so that I can see what you're trying to achieve?

I wish I could. Unfortunately it is closed source. Basically, I let the CMake android gradle plugin build a static library, then I pass extra environment variables with paths to those static libraries in the `exec` block and consume them in my `build.rs`.

> In general, I am happy to turn rust-android-gradle into a glue layer between Gradle and invoking cargo in various ways. Provided, of course, that we keep the main functionality (build, maybe check, maybe test) supported by default, preferably with no configuration required.

Taking this one step further: one of the ergonomic issues with Rust development for Android is that you cannot invoke `cargo` (except indirected through Gradle) since you are missing the appropriate environment variables. In my case, `cargo ndk` isn't enough to bridge the gap, since I would lose `exec`-based customization.

One way to address this would be to abandon the "commands" direction and instead have a per-(productFlavor,buildType,ABI) environment file. Instead of (or perhaps, in addition to) directly setting environment variables, Rust Android Gradle could write to this file (could be placed in `$module/env/$productFlavor/$buildType/$target_triple`), and then a specialized crate (called e.g. `rust-android-gradle`) could read and load the environment in a `build.rs` file.

The target triple, product flavor, and build type can be read via [environment variables in `build.rs`](https://doc.rust-lang.org/cargo/reference/environment-variables.html#environment-variables-cargo-sets-for-build-scripts). The target triple is available via `TARGET`. We can specify product flavor and build type via non-default package features (`CARGO_FEATURE_*`). This lets `rust-android-gradle` know where to lookup the environment file to use.

Example: assume you have product flavors "demo" and "full" and build types "alpha" and "beta". If you wanted to use `cargo` via CLI rather than Gradle, you would issue commands like this from inside `module`:

```bash
RUST_ANDROID_GRADLE_BUILD_ENV=demo+beta cargo build --target aarch64-linux-android
```

This is getting complex, though, so I'm not totally convinced of this direction. Also, TIL flavors can have [multiple dimensions](https://developer.android.com/studio/build/build-variants.html#flavor-dimensions) so you can have a lot more build variants:

> The number of build variants Gradle creates is equal to the product of the number of flavors in each flavor dimension and the number of build types you configure.

e.g. `myTask${flavorDim1}${flavorDim2}...${buildType}${ABI}`
